### ISSUE TITLE
LLVM 11 optimizations breaks code

### ISSUE THREAD

**Issue by bcardiff** [crystal-lang/crystal#10220]
After #9829 I attempt to bump homebrew formula with a backport patch. This will allow broader testing of llvm 11 for Crystal.
But the story didn't end well.

Similar to the issue https://github.com/crystal-lang/crystal/pull/9829#issuecomment-749183975 , I found another small program that will produce SIGILL / ud2 instructions.

```cr
# file: bug.ign.cr
require "yaml"

data = YAML.parse("version: 1.0\nshards:\n  molinillo:\n    github: crystal-lang/crystal-molinillo\n    version: 0.1.0\n\n")
puts data.raw.as(Hash(YAML::Any, YAML::Any)).keys.size
```

This program fails to run when compiled in `--release`. 

Part of the output was

```asm
    0x100034654 <+9540>: jmp    0x10003466c               ; <+9564> at parser.cr
    0x100034656 <+9542>: leaq   0x25e83(%rip), %rdi
    0x10003465d <+9549>: movq   %rbp, %rsi
    0x100034660 <+9552>: jmp    0x10003466c               ; <+9564> at parser.cr
    0x100034662 <+9554>: leaq   0x26047(%rip), %rdi
    0x100034669 <+9561>: movq   %r12, %rsi
    0x10003466c <+9564>: movq   %rbx, %rdx
    0x10003466f <+9567>: callq  0x10003a390               ; new at yaml.cr:86
    0x100034674 <+9572>: movq   %rax, %rdi
    0x100034677 <+9575>: callq  0x100031f60               ; raise at raise.cr:191
    0x10003467c <+9580>: leaq   0x2602d(%rip), %rdi
    0x100034683 <+9587>: jmp    0x10003465d               ; <+9549> at parser.cr
    0x100034685 <+9589>: leaq   0x26314(%rip), %rdi
    0x10003468c <+9596>: jmp    0x10003465d               ; <+9549> at parser.cr
    0x10003468e <+9598>: movq   %r12, %rdi
    0x100034691 <+9601>: callq  0x10003ad20               ; parse_float? at core.cr:324
->  0x100034696 <+9606>: ud2 
```

```cr
    case string
    when .starts_with?("0x"),
         .starts_with?("+0x"),
         .starts_with?("-0x")
      value = string.to_i64?(base: 16, prefix: true)
      value || string
    when .starts_with?("0."),
         .starts_with?('.')
      value = parse_float?(string) # <- this callq is followed by a ud2
      value || string
```

The similar structure that I can see between this case and the one in https://github.com/crystal-lang/crystal/pull/9829#issuecomment-749183975  is that:

1. the return value of the method is determined directly by the case when statements
2. the method called in the when returns a union 
3. the case returns another type of union that involves even more types.

But I still think there is an llvm issue here. But I could not reduce it further. What I did was:

1. Build crystal with llvm 11
2. Compiler above sample to .ll: `bin/crystal build --no-debug --emit=llvm-ir bug.ign.cr` (without --release)
3. Use llvm11 to compile and link the .ll without optimizations (-O0) and watch it run successfully
4. Use llvm11 to compile and link the same .ll with optimizations (-O3) and watch it crash with a due to the ud2
5. Use llvm10 to compile and link the same .ll with optimizations (-O3) and watch it run successfully.

So same .ll can be optimized and run on llvm10 but not in llvm11. That seems like an llvm issue to me.

To keep with this oddity, when used llvm11 -O1 I reach some memory corruption, because the program did not reach a ud2, but the yaml hash had less entries...

---

For reference there are aprox the commands I used to go back and forth with this issue.

```
$ bin/crystal build --no-debug --emit=llvm-ir bug.ign.cr

# commands to optimise, compile, link that I used.

$ opt -O3 -S input.ll > program.ll
$ llc program.ll -filetype=obj -o program.o
$ clang program.o -o program -rdynamic -L/nix/store/k2vpz20zhnr21zlxa0856qp1mir13b7j-libyaml-0.2.5/lib -lyaml -lpcre -lgc -lpthread /Users/bcardiff/Projects/crystal/master/src/ext/libcrystal.a -L/nix/store/1jafrk546ikjc6rcvc40pjhxxba3hkfd-libevent-2.1.12/lib -levent -liconv -ldl
$ ./program
```

**Comment by Fryguy** [comment]
@bcardiff I wonder if it would be useful, in general, to have a list of "known bad", or "known problematic" versions of LLVM, and then if the compiler detects that version it spits out a warning, perhaps with specifics on what to do.  Not sure _exactly_ how this would play out, but wanted to get idea to you.

Thinking that might help mitigate questions like https://github.com/crystal-lang/crystal/issues/9765 .  For me personally, at least, I recently spent a few hours trying to fix broken builds before finding out it was a known problem with LLVM 10, and that time would have been saved if the compiler just told me that LLVM 10 on Mac is known to be problematic in the particular way I was seeing.

**Comment by asterite** [comment]
To me the solution is to ship LLVM with the compiler at a fixed version. Relying on external versions is always problematic and we have to play the turtle and rabbit race, and we are always the turtle.

If we make sure LLVM 9 is shipped with the compiler then problem solved: there's no rush to fix this for LLVM 11, we can take our time to fix it, and we can release 1.0 with LLVM 9. After all, LLVM 11 doesn't bring any new features that Crystal uses.

**Comment by bcardiff** [comment]
To clarify: llvm 10 is fine, the problem was that in homebrew is no longer available and the name started pointing to llvm 11. Currently crystal in brew switched back to 9 as a workaround.

It's not good to stay out of date with llvm. 

Either way removing old llvm version support and sticking to one is another story. This is about llvm 11 not playing nice for us yet. So I would rather submit the issue before 🔥 📄 in my working copy.

**Comment by asterite** [comment]
> It's not good to stay out of date with llvm

Why not?

**Comment by jhass** [comment]
Because we cannot assume we have full control over binary packages, most distributions will want to ship their own from source builds. Then many discourage or forbid vendoring in dependencies like this. Finally we would become a huge pain to package maintainers if we force them to keep support old versions of a project as big as LLVM.

**Comment by asterite** [comment]
I'll be reducing the code in this issue to see if I can understand what's going on. Most likely we are doing something wrong.

**Comment by bcardiff** [comment]
What I concluded last time was that 

1. When we output a .ll file, that file will have some constructs that failed on sroa optimization. Odly, it seems that memory representation are fine.



```ll
; ModuleID = 'main_module'
source_filename = "main_module"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx"

%String = type { i32, i32, i32, i8 }
%"Array(String)" = type { i32, i32, i32, i32, %String** }
%"Exception::CallStack" = type { %"Array(Pointer(Void))"*, %"Array(String)"* }
%"Array(Pointer(Void))" = type { i32, i32, i32, i32, i8** }

; Function Attrs: uwtable
define i32 @main(i32 %argc, i8** %argv) #0 {
alloca:
  %0 = alloca %"Exception::CallStack", align 8
  br label %entry

entry:                                            ; preds = %alloca
  %1 = bitcast %"Exception::CallStack"* %0 to i8*
  call void @llvm.memset.p0i8.i64(i8* align 4 %1, i8 0, i64 mul nuw (i64 ptrtoint (i1** getelementptr (i1*, i1** null, i32 1) to i64), i64 2), i1 false)
  ret i32 0
}

declare void @llvm.memset.p0i8.i64(i8* nocapture writeonly, i8, i64, i1 immarg) #2

attributes #0 = { uwtable "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf"="true" }
attributes #2 = { argmemonly nounwind willreturn writeonly }
``` 

That file will fail in `opt -sroa -instcombine -simplifycfg -sroa -S input.ll > program.ll`

but changing that with the following equivalent seems to work,

```diff
+++ %2 = getelementptr i1*, i1** null, i32 1
+++ %3 = ptrtoint i1** %2 to i64
+++ %4 = mul nuw i64 %3, 2
+++ call void @llvm.memset.p0i8.i64(i8* align 4 %1, i8 0, i64 %4, i1 false)
--- call void @llvm.memset.p0i8.i64(i8* align 4 %1, i8 0, i64 mul nuw (i64 ptrtoint (i1** getelementptr (i1*, i1** null, i32 1) to i64), i64 2), i1 false)
```

I think this is not new to llvm@11 but it got in the middle of analyzing the underlying issue.

I couldn't find a way to generate that same ll from crystal directly. I couldn't find how to force those terms into variable, unless I use alloca.

```diff
# file /src/compiler/crystal/codegen/codegen.cr
    def allocate_aggregate(type)
      struct_type = llvm_struct_type(type)
      if type.passed_by_value?
        @last = alloca struct_type
      else
        if type.is_a?(InstanceVarContainer) && !type.struct? &&
           type.all_instance_vars.each_value.any? &.type.has_inner_pointers?
          @last = malloc struct_type
        else
          @last = malloc_atomic struct_type
        end
      end
+++   size = alloca @llvm_context.int64, "size"
+++   store struct_type.size, size
+++   memset @last, int8(0), load(size)
---   memset @last, int8(0), struct_type.size
      type_ptr = @last
      run_instance_vars_initializers(type, type, type_ptr)
      @last = type_ptr
    end
```

At this point I needed to reduce again the optimizations to find which one is breaking the code. These phases can be listed by `$ llvm-as < /dev/null | opt -O1 -disable-output -debug-pass=Arguments`

**Comment by asterite** [comment]
What I'm trying to do is to reduce this code:

```crystal
require "yaml"

data = YAML.parse("version: 1.0\nshards:\n  molinillo:\n    github: crystal-lang/crystal-molinillo\n    version: 0.1.0\n\n")
puts data.raw.as(Hash(YAML::Any, YAML::Any)).keys.size
```

to something much simpler that will reveal the issue.

**Comment by asterite** [comment]
With reducing the code I mean... this code also triggers the illegal instruction:

```crystal
enum EventKind
  SCALAR
  MAPPING_START
  MAPPING_END
end

struct Any
  alias Type = Float64 | String | Time | Hash(Any, Any)

  def initialize(@raw : Type)
  end
end

class Parser
  def initialize
    @pull_parser = PullParser.new
  end

  def parse
    if @pull_parser.kind.scalar?
      value = Any.new(parse_scalar(@pull_parser.value))
      @pull_parser.read_next
      return value
    end

    @pull_parser.read_next

    raw = {} of Any => Any
    until @pull_parser.kind.mapping_end?
      raw[parse] = parse
    end
    Any.new(raw)
  end

  def parse_scalar(string)
    if string.starts_with?("0.")
      return string.to_f64? || string
    end

    if string[0].ascii_number? && string.size >= 8
      return Time.local
    end

    string
  end
end

class PullParser
  record Event, kind : EventKind, value : String

  def initialize
    @events = [
      Event.new(:mapping_start, ""),
      Event.new(:scalar, "version"),
      Event.new(:scalar, "1.0"),
      Event.new(:scalar, "shards"),
      Event.new(:mapping_start, ""),
      Event.new(:scalar, "molinillo"),
      Event.new(:mapping_start, ""),
      Event.new(:scalar, "github"),
      Event.new(:scalar, "crystal-lang/crystal-molinillo"),
      Event.new(:scalar, "version"),
      Event.new(:scalar, "0.1"),
      Event.new(:mapping_end, ""),
      Event.new(:mapping_end, ""),
      Event.new(:mapping_end, ""),
    ] of Event
    @index = -1

    read_next
  end

  def kind
    @events[@index].kind
  end

  def value
    @events[@index].value
  end

  def read_next
    @index += 1
  end
end

data = Parser.new.parse
puts data
data.@raw.as(Hash).keys
```

but it's much easier than the original one, which depends on the entire YAML library.

I will keep reducing the code until I find the smallest code that reproduces the issue... though I'm starting to suspect it has something to do with parsing floats.

**Comment by asterite** [comment]
This is the power of reductions. This code reproduces the issue:

```crystal
def foo
  ("".to_f64? || "").as(Float64 | String | Time)
end

puts foo
```

I'll try to reduce it even further so it doesn't depend on `to_f64?` (maybe the problem is there).

**Comment by asterite** [comment]
"Reduced" a bit more:

```crystal
def foo
  (bar || "").as(Float64 | String | UInt8[24])
end

def bar
  endptr = Pointer(UInt8).malloc(1_u64)
  string_end = Pointer(UInt8).malloc(1_u64)

  return if endptr == string_end

  endptr < endptr

  1.0
end

foo
```

It's strange because the code above doesn't compile with `--prelude=empty`: it's missing the `Pointer#==` and `Pointer#<` methods. But when I define those it doesn't fail anymore. So there must be something else going on...

**Comment by asterite** [comment]
A bit more:

```crystal
def foo
  (bar || "").as(Float64 | String | UInt8[24])
end

def bar
  ptr = Pointer(UInt8).malloc(1_u64)
  ptr.address == ptr.address + 1 ? nil : 1.0
end

foo
```

I'm starting to suspect casting between different union types and that somehow corrupting the stack. The strange thing is that LLVM would have to notice this and then replace the entire program with `unreachable`, because when that code is optimized that's what happens...

**Comment by asterite** [comment]
Another reduction:

```crystal
def foo
  (bar || true).as(Int32 | Bool | UInt8[12])
end

def bar
  ptr = Pointer(UInt8).malloc(1_u64)
  ptr2 = Pointer(UInt8).malloc(1_u64)
  ptr.address == ptr2.address ? nil : 1
end

foo
```

This time no strings are involved. And that program actually compiles with `--prelude=empty` but in that case it doesn't crash 🤷
